internal/provider/resource_password_test.go

```diff
@@ -1153,64 +1166,102 @@ func TestUpgradePasswordStateV0toV3(t *testing.T) {
-		Lower:           types.Bool{Value: true},
-		Number:          types.Bool{Value: true},
-		Numeric:         types.Bool{Value: true},
-		MinNumeric:      types.Int64{Value: 0},
-		MinUpper:        types.Int64{Value: 0},
-		MinLower:        types.Int64{Value: 0},
-		MinSpecial:      types.Int64{Value: 0},
-		OverrideSpecial: types.String{Value: "!#$%\u0026*()-_=+[]{}\u003c\u003e:?"},
-		Result:          types.String{Value: "DZy_3*tnonj%Q%Yx"},
+	expectedResp := &res.UpgradeStateResponse{
+		State: tfsdk.State{
+			Raw: tftypes.NewValue(tftypes.Object{
+				AttributeTypes: map[string]tftypes.Type{
+					"bcrypt_hash":      tftypes.String,
+					"id":               tftypes.String,
+					"keepers":          tftypes.Map{ElementType: tftypes.String},
+					"length":           tftypes.Number,
+					"lower":            tftypes.Bool,
+					"min_lower":        tftypes.Number,
+					"min_numeric":      tftypes.Number,
+					"min_special":      tftypes.Number,
+					"min_upper":        tftypes.Number,
+					"number":           tftypes.Bool,
+					"numeric":          tftypes.Bool,
+					"override_special": tftypes.String,
+					"result":           tftypes.String,
+					"special":          tftypes.Bool,
+					"upper":            tftypes.Bool,
+				},
+			}, map[string]tftypes.Value{
+				"bcrypt_hash":      tftypes.NewValue(tftypes.String, "hash"),
+				"id":               tftypes.NewValue(tftypes.String, "none"),
+				"keepers":          tftypes.NewValue(tftypes.Map{ElementType: tftypes.String}, nil),
+				"length":           tftypes.NewValue(tftypes.Number, 16),
+				"lower":            tftypes.NewValue(tftypes.Bool, true),
+				"min_lower":        tftypes.NewValue(tftypes.Number, 0),
+				"min_numeric":      tftypes.NewValue(tftypes.Number, 0),
+				"min_special":      tftypes.NewValue(tftypes.Number, 0),
+				"min_upper":        tftypes.NewValue(tftypes.Number, 0),
+				"number":           tftypes.NewValue(tftypes.Bool, true),
+				"numeric":          tftypes.NewValue(tftypes.Bool, true),
+				"override_special": tftypes.NewValue(tftypes.String, "!#$%\u0026*()-_=+[]{}\u003c\u003e:?"),
+				"result":           tftypes.NewValue(tftypes.String, "DZy_3*tnonj%Q%Yx"),
+				"special":          tftypes.NewValue(tftypes.Bool, true),
+				"upper":            tftypes.NewValue(tftypes.Bool, true),
+			}),
+			Schema: passwordSchemaV3(),
+		},
 	}
 
-	actual := passwordModelV3{}
-	diags := resp.State.Get(context.Background(), &actual)
-	if diags.HasError() {
-		t.Errorf("error getting state: %v", diags)
-	}
+	// rawTransformed allows equality testing to be used by mutating the bcrypt_hash value in the response to a known value.
+	rawTransformed, err := tftypes.Transform(resp.State.Raw, func(path *tftypes.AttributePath, value tftypes.Value) (tftypes.Value, error) {
+		bcryptHashPath := tftypes.NewAttributePath().WithAttributeName("bcrypt_hash")
 
-	err := bcrypt.CompareHashAndPassword([]byte(actual.BcryptHash.Value), []byte(actual.Result.Value))
```

Comment:
Can we keep the `bcrypt.CompareHashAndPassword()` testing to ensure we're generating valid bcrypt hashes before the attribute is removed for the equality testing?

Reply:
My bad. I neglected to reinstate these. Thanks for pointing this out.